File: iphone-code/code/Stasher/STParentPaymentFlowViewController.m
//
//  STParentPaymentFlowViewController.m
//  Stasher
//
//  Created by bhushan on 06/02/15.
//  Copyright (c) 2015 OAB. All rights reserved.
//

#import "STParentPaymentFlowViewController.h"

@interface STParentPaymentFlowViewController ()

@end

@implementation STParentPaymentFlowViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    [self.headerLabel setFont:[UIFont FontForHeader]];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(bankTransactionFinished:) name:kNotificationIdentifier_BankTransactionFinished object:nil];
    //create dictionary with user information
    NSMutableDictionary *userInfo = [[NSMutableDictionary alloc] init];
    
    //add api key
    [userInfo setObject:@"4aa327f5ac661878d6178e2c2b44aa5a9a0708d2" forKey:@"api_key"];
    
    //add api password
    [userInfo setObject:@"7a383e494ac323569402b9f1fe591c1735906202" forKey:@"api_password"];
    
    //add amount as string
    if (self.paymentParamsDict && [self.paymentParamsDict objectForKey:@"amount"]!=nil) {
        [userInfo setObject:[self.paymentParamsDict objectForKey:@"amount"] forKey:@"amount"];
    }
    //[userInfo setObject:@"0.01" forKey:@"amount"];
    
    //add invoice detail];
    if (self.paymentParamsDict && [self.paymentParamsDict objectForKey:@"description"]!=nil) {
        [userInfo setObject:[self.paymentParamsDict objectForKey:@"description"] forKey:@"invoice_detail"];
    }
    //[userInfo setObject:@"invoice detail" forKey:@"invoice_detail"];
    
    //add information request
    [userInfo setObject:@"show_all" forKey:@"information_request"];
    
    //add recurring detail
    [userInfo setObject:@"ot" forKey:@"reccuring"];
    
    //add redirect URL
    [userInfo setObject:@"fb743090835784071://" forKey:@"redirect_url"];
    
    //make url string with user info on the end
    NSString *url = [NSString stringWithFormat:@"https://knoxpayments.com/pay?api_key=%@&api_password=%@&amount=%@&invoice_detail=%@&information_request=%@&recurring=%@&redirect_url=%@",[userInfo objectForKey:@"api_key"],[userInfo objectForKey:@"api_password"],[userInfo objectForKey:@"amount"],[userInfo objectForKey:@"invoice_detail"],[userInfo objectForKey:@"information_request"],[userInfo objectForKey:@"reccuring"],[userInfo objectForKey:@"redirect_url"] ];
    url = [url stringByAddingPercentEscapesUsingEncoding:NSASCIIStringEncoding];
    NSLog(@"%@",url);
    
    //create URL and request
    NSURL *nsurl = [NSURL URLWithString:url];
    NSURLRequest *nsrequest = [NSURLRequest requestWithURL:nsurl];
    [self.paymentWebView loadRequest:nsrequest];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (IBAction)backButtonPressed:(id)sender
{
    [self.navigationController popViewControllerAnimated:YES];
}

- (void)bankTransactionFinished:(NSNotification*)notification
{
    NSDictionary *dict = [notification userInfo];
    if (dict) {
        if ([[dict objectForKey:@"completed"] isEqualToString:@"canceled"]) {
            [self.navigationController popViewControllerAnimated:YES];
            return;
        }
        if ([dict objectForKey:@"pay_id"]) {
            [[STUserIdentity sharedInstance] setUserKnoxFirstTransactionID:[dict objectForKey:@"pay_id"]];
            if (httpReq) {
                httpReq.delegate = nil;
                httpReq = nil;
            }
            httpReq = [[BOABHttpReq alloc] initBoabReqWithDelegate:self shouldShowActivityIndicatorView:YES];
           
            if (self.bankAccountTypeInt == 1) {
                NSString *apiActionString;
                apiActionString = kAPIActionAddBankAccount;
                [httpReq sendAsyncPostRequest:STASHER_BASE_URL params:[NSDictionary dictionaryWithObjectsAndKeys:[NSNumber numberWithInt:[[[STUserIdentity sharedInstance] userId] intValue]], kParamKeyParentID,apiActionString,kParamKeyAction,[dict objectForKey:@"pay_id"],kParamKeyTransactionID,nil] json:YES];
            }else if (self.bankAccountTypeInt == 2){
                NSString *apiActionString;
                apiActionString = kAPIActionAddBankAccountChild;
                NSString *childUserID;
                childUserID = [[self.paymentParamsDict objectForKey:@"childInfoDict"] objectForKey:kParamKeyUserID];
                [httpReq sendAsyncPostRequest:STASHER_BASE_URL params:[NSDictionary dictionaryWithObjectsAndKeys:[NSNumber numberWithInt:[childUserID intValue]], kParamKeyChildID,apiActionString,kParamKeyAction,[dict objectForKey:@"pay_id"],kParamKeyTransactionID,nil] json:YES];
            }
        }
        if ([dict objectForKey:@"completed"]) {
            
        }
    }
}

#pragma mark ----- WebView Delegate

- (void)webViewDidStartLoad:(UIWebView *)webView
{
    if (self.activityIndicatorView!=nil) {
        [self.activityIndicatorView removeFromSuperview];
        self.activityIndicatorView = nil;
    }
    self.activityIndicatorView = [[STActivityIndicatorView alloc] initWithDefaultSizeandSuperView:self.view];
    [self.activityIndicatorView startAnimation];
    
    
}
- (void)webViewDidFinishLoad:(UIWebView *)webView
{
    [self.activityIndicatorView stopAnimation];
}

- (void)webView:(UIWebView *)webView didFailLoadWithError:(NSError *)error
{
     [self.activityIndicatorView stopAnimation];
    if (self.activityIndicatorView!=nil) {
        [self.activityIndicatorView removeFromSuperview];
        self.activityIndicatorView = nil;
    }
}


#pragma mark ----- BOABHttpReqDelegates

- (void)BOABHttpReqFinishedWithResponse:(NSData*)resonseData andConnection:(BOABURLConnection*)conn
{
    if (resonseData) {
        NSDictionary *responseDictionary = [NSJSONSerialization JSONObjectWithData:resonseData options:kNilOptions error:nil];
        if (responseDictionary) {
            if ([self.delegate respondsToSelector:@selector(addAccountSuccessFullWithDict:andBankAccountUserType:)]) {
                [self.delegate addAccountSuccessFullWithDict:responseDictionary andBankAccountUserType:self.bankAccountTypeInt];
            }
        }
    }
}


@end

File: swift-code/Stasher/STParentPaymentFlowViewController.swift
//  Converted with Swiftify v1.0.6292 - https://objectivec2swift.com/
//
//  STParentPaymentFlowViewController.swift
//  Stasher
//
//  Created by bhushan on 06/02/15.
//  Copyright (c) 2015 OAB. All rights reserved.
//
import UIKit
protocol ParentPaymentFlowDelegate: NSObjectProtocol {
    func addAccountSuccessFull(withDict dict: [AnyHashable: Any], andBankAccountUserType bankAccType: Int)
}
class STParentPaymentFlowViewController: UIViewController, BOABHttpReqDelegate {
    var httpReq: BOABHttpReq?

    weak var delegate: ParentPaymentFlowDelegate?
    @IBOutlet var headerLabel: UILabel!
    @IBOutlet var paymentWebView: UIWebView!
    var paymentParamsDict = [AnyHashable: Any]()
    var activityIndicatorView: STActivityIndicatorView?
    var bankAccountTypeInt: Int = 0


    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        headerLabel.font = UIFont()
        NotificationCenter.default.addObserver(self, selector: #selector(self.bankTransactionFinished), name: kNotificationIdentifier_BankTransactionFinished, object: nil)
            //create dictionary with user information
        var userInfo = [AnyHashable: Any]()
        //add api key
        userInfo["api_key"] = "4aa327f5ac661878d6178e2c2b44aa5a9a0708d2"
        //add api password
        userInfo["api_password"] = "7a383e494ac323569402b9f1fe591c1735906202"
        //add amount as string
        if paymentParamsDict && paymentParamsDict["amount"] != nil {
            userInfo["amount"] = paymentParamsDict["amount"]
        }
        //[userInfo setObject:@"0.01" forKey:@"amount"];
        //add invoice detail];
        if paymentParamsDict && paymentParamsDict["description"] != nil {
            userInfo["invoice_detail"] = paymentParamsDict["description"]
        }
        //[userInfo setObject:@"invoice detail" forKey:@"invoice_detail"];
        //add information request
        userInfo["information_request"] = "show_all"
        //add recurring detail
        userInfo["reccuring"] = "ot"
        //add redirect URL
        userInfo["redirect_url"] = "fb743090835784071://"
            //make url string with user info on the end
        var url: String? = "https://knoxpayments.com/pay?api_key=\(userInfo["api_key"] as? String)&api_password=\(userInfo["api_password"] as? String)&amount=\(userInfo["amount"] as? String)&invoice_detail=\(userInfo["invoice_detail"] as? String)&information_request=\(userInfo["information_request"] as? String)&recurring=\(userInfo["reccuring"] as? String)&redirect_url=\(userInfo["redirect_url"] as? String)"
        url = url?.addingPercentEscapes(using: String.Encoding.ascii)
        print("\(url)")
            //create URL and request
        var nsurl = URL(string: url)
        var nsrequest = URLRequest(url: nsurl!)
        paymentWebView.loadRequest(nsrequest)
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    @IBAction func backButtonPressed(_ sender: Any) {
        navigationController?.popViewController(animated: true)
    }

    func bankTransactionFinished(_ notification: Notification) {
        var dict: [AnyHashable: Any]? = notification.userInfo
        if dict != nil {
            if (dict?["completed"] == "canceled") {
                navigationController?.popViewController(animated: true)
                return
            }
            if dict?["pay_id"] {
                STUserIdentity.sharedInstance().userKnoxFirstTransactionID = dict?["pay_id"]
                if httpReq != nil {
                    httpReq?.delegate = nil
                    httpReq = nil
                }
                httpReq = BOABHttpReq.initBoabReq(withDelegate: self, shouldShowActivityIndicatorView: true)
                if bankAccountTypeInt == 1 {
                    var apiActionString: String
                    apiActionString = kAPIActionAddBankAccount
                    httpReq?.sendAsyncPostRequest(STASHER_BASE_URL, params: [
                        kParamKeyParentID : Int(CInt(STUserIdentity.sharedInstance().userId())),
                        kParamKeyAction : apiActionString,
                        kParamKeyTransactionID : dict?["pay_id"]
                    ]
, json: true)
                }
                else if bankAccountTypeInt == 2 {
                    var apiActionString: String
                    apiActionString = kAPIActionAddBankAccountChild
                    var childUserID: String
                    childUserID = (paymentParamsDict["childInfoDict"][kParamKeyUserID] as? String)
                    httpReq?.sendAsyncPostRequest(STASHER_BASE_URL, params: [
                        kParamKeyChildID : Int((childUserID as NSString).integerValue),
                        kParamKeyAction : apiActionString,
                        kParamKeyTransactionID : dict?["pay_id"]
                    ]
, json: true)
                }
            }
            if dict?["completed"] {

            }
        }
    }
// MARK: ----- WebView Delegate

    func webViewDidStartLoad(_ webView: UIWebView) {
        if activityIndicatorView != nil {
            activityIndicatorView?.removeFromSuperview()
            activityIndicatorView = nil
        }
        activityIndicatorView = STActivityIndicatorView(defaultSizeandSuperView: view)
        activityIndicatorView?.startAnimation()
    }

    func webViewDidFinishLoad(_ webView: UIWebView) {
        activityIndicatorView?.stopAnimation()
    }

    func webView(_ webView: UIWebView, didFailLoadWithError error: Error) {
        activityIndicatorView?.stopAnimation()
        if activityIndicatorView != nil {
            activityIndicatorView?.removeFromSuperview()
            activityIndicatorView = nil
        }
    }
// MARK: ----- BOABHttpReqDelegates

    func boabHttpReqFinished(withResponse resonseData: Data, andConnection conn: BOABURLConnection) {
        if resonseData {
            var responseDictionary: [AnyHashable: Any]? = try? JSONSerialization.jsonObject(withData: resonseData, options: kNilOptions)
            if responseDictionary != nil {
                if delegate?.responds(to: Selector("addAccountSuccessFullWithDict:andBankAccountUserType:")) {
                    delegate?.addAccountSuccessFull(withDict: responseDictionary, andBankAccountUserType: bankAccountTypeInt)
                }
            }
        }
    }
}